test/test_torch.py

```diff
@@ -16706,6 +16417,216 @@ def caller(cls,
 
+def _generic_input_fn(dtype, device):
+    input = []
+    input.append(list(range(-5, 5)))
+    input.append([0 for x in range(-5, 5)])
+    input.append([x + 1e-6 for x in range(-5, 5)])
+    # Some vectorized implementations don't support large ranges
+    input.append([x + 1e10 for x in range(-5, 5)])
+    input.append([x - 1e10 for x in range(-5, 5)])
+    input.append(torch.randn(10).tolist())
+    input.append((torch.randn(10) + 1e6).tolist())
+    input.append([math.pi * (x / 2) for x in range(-5, 5)])
+    return torch.tensor(input, dtype=dtype, device=device)
+
+def _gamma_input_fn(dtype, device, test_poles=True):
+    input = []
+    input.append((torch.randn(10).abs() + 1e-4).tolist())
+    input.append((torch.randn(10).abs() + 1e6).tolist())
+    zeros = torch.linspace(-9.5, -0.5, 10)
+    input.append(zeros.tolist())
+    input.append((zeros - 0.49).tolist())
+    input.append((zeros + 0.49).tolist())
+    input.append((zeros + (torch.rand(10) * 0.99) - 0.5).tolist())
+    if test_poles:
+        input.append([-0.999999994, -1.999999994, -2.0000000111,
+                      -100.99999994, -1931.99999994, 0.000000111,
+                      -0.000000111, 0, -2, -329])
+    return torch.tensor(input, dtype=dtype, device=device)
+
+class _TorchTestMeta(object):
+    def __init__(self,
+                 opstr,
+                 args=[],
+                 reffn=None,
+                 refargs=[],
+                 input_fn=_generic_input_fn,
+                 inputargs=[],
+                 substr=None,
+                 make_inplace=True,
+                 decorators=[onlyCPU],
+                 ref_decorator='numpy',
+                 rtol=None,
+                 atol=None,
+                 dtypes=_float_types_no_half,
+                 replace_inf_with_nan=False):
+        self.opstr = opstr
+        self.args = args
+        self.reffn = reffn
+        if refargs == []:
+            self.refargs = lambda x: (x.numpy(),)
+        else:
+            self.refargs = refargs
+        self.input_fn = input_fn
+        self.inputargs = inputargs
+        self.substr = substr if substr is not None else ''
+        self.make_inplace = make_inplace
+        assert ref_decorator == 'numpy' or ref_decorator == 'scipy'
+        if ref_decorator == 'numpy':
+            self.ref_decorator = [unittest.skipIf(not TEST_NUMPY, "Numpy not found")]
+        elif ref_decorator == 'scipy':
+            self.ref_decorator = [unittest.skipIf(not TEST_SCIPY, "Scipy not found")]
+        self.decorators = decorators
+        self.rtol = rtol
+        self.atol = atol
+        self.dtypes = dtypes
+        self.replace_inf_with_nan = replace_inf_with_nan
+
+torch_op_tests = [_TorchTestMeta('sin'),
+                  _TorchTestMeta('sinh'),
+                  _TorchTestMeta('lgamma', reffn=scipy.special.gammaln, ref_decorator='scipy', make_inplace=False),
+                  _TorchTestMeta('asin', reffn=np.arcsin, make_inplace=False),
```

Comment:
Why do we have a reference for asin but not sin?

Reply:
I see now that reffn is only required to be specified if there's a name mismatch. My mistake!